### ISSUE TITLE
Loading .npy images instead of the normal image formats (jpg,png...etc)

### ISSUE THREAD

**Issue by LilNader** [junyanz/pytorch-CycleGAN-and-pix2pix#500]
Dears 

I need to load numpy data directly to the neural network without any transformations applied, where the Numpy data has the same shape as the images.

What is the simplest approach for this implementation and where exactly to apply the changes in the code?

Note: I have organized the Numpy data in the same folder structure of the images (trainA, trainB,..etc.)
 

Thank you

**Comment by junyanz** [comment]
You can change these two [lines](https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix/blob/master/data/unaligned_dataset.py#L57).  Also see the template dataset [class](https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix/blob/master/data/template_dataset.py) on how to implement a custom class.

**Comment by LilNader** [comment]
> You can change these two [lines](https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix/blob/master/data/unaligned_dataset.py#L57). Also see the template dataset [class](https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix/blob/master/data/template_dataset.py) on how to implement a custom class.

Thanks very much :)

**Comment by joeljoeljo** [comment]
@LilNader  Are your datasets in .npy format? May i know what did u change your lines to. ty!

**Comment by LilNader** [comment]
> @LilNader May i know what did u change your lines to. ty!
I used the template data class as mentioned by @junyanz then I changed the [two lines](https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix/blob/master/data/unaligned_dataset.py#L57) as mentioned by  to:
A_img = np.load(A_path)
B_img = np.load(B_path)

In my problem I set preprocess to none
I also made a small change in base_dataset.py
- I replaced this [line](https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix/blob/18a40e606eb5ef5214db84b1bb24b9f0e3641371/data/base_dataset.py#L98) by continue

**Comment by villanuevab** [comment]
When I do this, the values seem to load in fine, but the outputs are all-white. Is there some post-processing I need to do before saving the images as .png? 

As an aside, I normalize the inputs in __getitem__ since the raw input data is not normalized. Could this be affecting training?

**Comment by patrick-han** [comment]
@LilNader I'm curious, were you working with 3-channel dumpy data (3 layers) or 4? I am currently trying to directly pass 4 layer/channels arrays but am having lots of trouble modifying the code since the PIL library seems to bank on 3 channels

**Comment by LilNader** [comment]
> @LilNader I'm curious, were you working with 3-channel dumpy data (3 layers) or 4? I am currently trying to directly pass 4 layer/channels arrays but am having lots of trouble modifying the code since the PIL library seems to bank on 3 channels

@patrick-han In my case I was passing single channel numpy data (binary images) directly to the Network without using PIL
In your case, I guess you will need to indicate the number of input and output channels using --input_nc or --output_nc

**Comment by yaxinnn** [comment]
@SirNader May I ask what do you mean by 'used the template data class'? The two changed lines are in the script called 'unaligned_dataset.py', but the template data class is called 'template_dataset.py'. When specifying the option of '--dataset_mode', which one did you use? 'unaligned' or template? Thank you in advance!

**Comment by povolann** [comment]
> > @LilNader I'm curious, were you working with 3-channel dumpy data (3 layers) or 4? I am currently trying to directly pass 4 layer/channels arrays but am having lots of trouble modifying the code since the PIL library seems to bank on 3 channels
> 
> @patrick-han In my case I was passing single channel numpy data (binary images) directly to the Network without using PIL and I had to set --input_nc and --output_nc to 1
> In your case, I guess you will need to indicate the number of input and output channels using --input_nc or --output_nc

Hello, I have tried to same as you (load npy files after all the changes you wrote here), but it doesn't really work for me.
So, I have used instead just np.load()
        A_img = Image.fromarray(np.load(A_path))
        B_img = Image.fromarray(np.load(B_path))
and in base_dataset.py I have commented this part (my images are already grayscale)
def get_transform(opt, params=None, grayscale=False, method=Image.BICUBIC, convert=True):
    transform_list = []
    # if grayscale:
    #     transform_list.append(transforms.Grayscale(1))

So, now the code works, but my images are super dark. And also, the results are still in png. So, I was wondering if you managed to change the saving of results to npy? Because maybe that would help me... After all, my npy files are type float32 and the png files are uint8, so there is probably some loss of precision....?
Thank you for any advice!
